RFC: Add an RFC Process (#1)
OpenDreamProject/rfcs: README.md

## Motivation

Although the core goal of OpenDream is to provide one-way parity with BYOND, there are [many opportunities](https://github.com/OpenDreamProject/OpenDream/wiki/Differences-Between-OpenDream-and-BYOND#enhancements) in which OpenDream and the DM language can be improved while still enabling users to migrate from BYOND with minimal effort. Or to put it another way, we can add new functionality *on top* of BYOND parity so that OpenDream users can take advantage of them while BYOND users are not impacted. 

A key example of this is [pragmas](https://github.com/OpenDreamProject/OpenDream/wiki/Pragmas---Error-Emissions). Pragmas provide configurable compiler emissions for things such as runtime errors that *can* be detected at compiletime but BYOND fails to detect. Utilizing pragmas while maintaining BYOND compatibility is as simple as declaring them in their own code file and putting this in your code somewhere:
```
#ifdef OPENDREAM

> #include("/path/to/pragma/file.dm") // Don't include the file of pragmas unless `OPENDREAM` is defined (which OpenDream defines for you)

Author: ```suggestion
// Don't include the file of pragmas unless `OPENDREAM` is defined (which OpenDream defines for you)
#include("/path/to/pragma/file.dm")
```

# OpenDream RFC Repo

> This repository stores RFCs for the [OpenDream](https://github.com/OpenDreamProject/OpenDream) project. Current and future users of OpenDream are strongly encouraged to watch this repository and provide constructive input on all open pull requests.

Reviewer: ```suggestion
This repository stores RFCs _(Request For Comments)_ for the [OpenDream](https://github.com/OpenDreamProject/OpenDream) project. Current and future users of OpenDream are strongly encouraged to watch this repository and provide constructive input on all open pull requests.
```

## Motivation

Although the core goal of OpenDream is to provide one-way parity with BYOND, there are [many opportunities](https://github.com/OpenDreamProject/OpenDream/wiki/Differences-Between-OpenDream-and-BYOND#enhancements) in which OpenDream and the DM language can be improved while still enabling users to migrate from BYOND with minimal effort. Or to put it another way, we can add new functionality *on top* of BYOND parity so that OpenDream users can take advantage of them while BYOND users are not impacted. 

A key example of this is [pragmas](https://github.com/OpenDreamProject/OpenDream/wiki/Pragmas---Error-Emissions). Pragmas provide configurable compiler emissions for things such as runtime errors that *can* be detected at compiletime but BYOND fails to detect. Utilizing pragmas while maintaining BYOND compatibility is as simple as declaring them in their own code file and putting this in your code somewhere:
```
#ifdef OPENDREAM
// Don't include the file of pragmas unless `OPENDREAM` is defined (which OpenDream defines for you)
#include("/path/to/pragma/file.dm")
#endif
```

> Furthermore, OpenDream is intended to provide a solution for *everyone* and not any specific codebase. To prevent any one person from turning the project into their own new programming language, this RFC process exists to solicit input from the wider community on any major changes that would impact their current or future usage of OpenDream.

Reviewer: Should "major changes" be defined (e.g Breaking parity with byond) or is it a judgement thing?

I see the section below but that looks like it's supposed to be a "Questions to ask during the RFC" rather than a "when is an RFC necessary".

I've seen it several times before that someone puts a massive amount of effort into something they personally believe is a good change, and after it's PRed they find out it's not something that's wanted. Seems like something we should be addressing sooner in the implementation process rather than "when someone decides it should be done"

Author: It's a judgement thing. People choosing to invest a lot of effort into something without asking if it's wanted is their problem; it never hurts to ask.

## RFC Considerations

Whenever new syntax is introduced, we need to ask:

- Is it BYOND compatible?
- Is it easy for machines and humans to parse?
- Does it create grammar ambiguities for current and future syntax?
- Is it stylistically coherent with the rest of the language?
- Does it present challenges with editor integration like autocomplete?

For changes in semantics, we should be asking:

- Is behavior easy to understand and non-surprising?
- Can it be implemented performantly today?
- Is it compatible with type checking and other forms of static analysis?

As the OpenDream userbase grows, reversing these decisions will only become more costly and have backwards compatibility implications.

## RFC Process

To open an RFC, a Pull Request must be opened which creates a new Markdown file in `docs/` folder. The RFCs should follow the template `TEMPLATE.md`, and should have a file name that matches the RFC's title.

> ### RFC Drafts
> If you are in the early stages of the RFC and would like to solicit feedback on major considerations for your proposal, but it isn't ready for formal review, you can open your RFC as a Draft Pull Request. Once you are ready, mark it as ready for review and proceed to the below section:

Author: ```suggestion
### RFC Drafts
If you are in the early stages of designing the RFC and would like to solicit feedback on major considerations for your proposal, but it isn't ready for formal review, you can open your RFC as a Draft Pull Request. Once you are ready, mark it as ready for review and proceed to the below section:
```

### RFC Review & Finalization
Every open RFC will be open for at least two calendar weeks. This is to make sure that there is sufficient time to review the proposal and raise concerns or suggest improvements. The discussion points should be reflected on the PR comments; when discussion happens outside of the comment stream, the points salient to the RFC should be summarized as a followup.

> The first week should be used to resolve any major technical or design problems raised by the OpenDream team, though everyone is encouraged to provide input. After the first week and any major problems are resolved, a member of the OpenDream team should share the RFC in the following places to solicit feedback from the wider community for the second week:

Reviewer 2: > ... raised by the OpenDream team

who is the OpenDream team? Maintainers on the repository?

Author: ```suggestion
The first week should be used to resolve any major technical or design problems raised by the OpenDream team, though everyone is encouraged to provide input. After the first week and any major problems are resolved, a Maintainer from the OpenDream team should share the RFC in the following places to solicit feedback from the wider community for the second week:
```

- `#codebase-announcements` channel in OpenDream's Discord.

> - `#headcoderbus` channel in coderbus's Discord.

Author: ```suggestion
- `#maintainerbus` channel in coderbus's Discord.
```

> - `#maintainerbus` channel in coderbus's Discord.

Reviewer 2: Links?

- `#coderbus` channel in coderbus's Discord.

When the initial comment period expires, the RFC can be merged if there's consensus that the change is important and that the details of the syntax/semantics presented are workable. The decision to merge the RFC is made by the OpenDream team.

> If at any point the RFC undergoes substantial changes (anything more than spelling/formatting/wording), the two-week timer should be reset to the time the last substantial commit was pushed.

Author: ```suggestion
If at any point the RFC undergoes substantial changes (anything more than spelling/formatting/wording), the two-week timer may (at OpenDream staff discretion) be reset to the time the last substantial commit was pushed.
```

> If at any point the RFC undergoes substantial changes (anything more than spelling/formatting/wording), the two-week timer may (at OpenDream staff discretion) be reset to the time the last substantial commit was pushed.

Reviewer 2: is `OpenDream staff` different than `team`?

Author: ```suggestion
If at any point the RFC undergoes substantial changes (anything more than spelling/formatting/wording), the two-week timer may (at OpenDream Maintainer discretion) be reset to the time the last substantial commit was pushed.
```
